array_analyzer/load/report.py

```diff
@@ -48,29 +55,48 @@ def write_antigen_report(writer, array_type):
         sheet = deepcopy(constants.WELL_OUTPUT_TEMPLATE)
         # loop all wells and write OD, INT, BG of this antigen
         if array_type == 'od':
-            for od_position, od_well in np.ndenumerate(constants.WELL_OD_ARRAY):
-                od_val = od_well[antigen_position[0], antigen_position[1]]
-                well_name = well_to_image[(od_position[0] + 1, od_position[1] + 1)]
-                sheet[well_name[0]][int(well_name[1:])] = od_val
+            sheet = write_to_sheet(sheet, constants.WELL_OD_ARRAY, antigen_position, well_to_image)
         elif array_type == 'int':
-            for int_position, int_well in np.ndenumerate(constants.WELL_INT_ARRAY):
-                int_val = int_well[antigen_position[0], antigen_position[1]]
-                well_name = well_to_image[(int_position[0] + 1, int_position[1] + 1)]
-                sheet[well_name[0]][int(well_name[1:])] = int_val
+            sheet = write_to_sheet(sheet, constants.WELL_INT_ARRAY, antigen_position, well_to_image)
         elif array_type == 'bg':
-            for bg_position, bg_well in np.ndenumerate(constants.WELL_BG_ARRAY):
-                bg_val = bg_well[antigen_position[0], antigen_position[1]]
-                well_name = well_to_image[(bg_position[0] + 1, bg_position[1] + 1)]
-                sheet[well_name[0]][int(well_name[1:])] = bg_val
+            sheet = write_to_sheet(sheet, constants.WELL_BG_ARRAY, antigen_position, well_to_image)
         else:
             raise AttributeError(f"report array type {array_type} not supported")
 
-        # write the outputs from the above three to a worksheet
+        # write the outputs from ONE of the above three to a worksheet
+        # (this function is called once for each OD, INT, BG)
         od_sheet_df = pd.DataFrame(sheet).T
 
+        sheet_name = f'{array_type}_{antigen_position[0]}_{antigen_position[1]}_{antigen}'
+        if len(sheet_name) >= 31:
+            warnings.warn("antigen sheet name is too long, truncating")
+            sheet_name = sheet_name[:31]
+
         od_sheet_df.to_excel(writer,
-                             sheet_name=f''
-                             f'{array_type}_'
-                             f'{antigen_position[0]}_'
-                             f'{antigen_position[1]}_'
-                             f'{antigen}')
+                             sheet_name=sheet_name)
+
+
+def write_to_sheet(sheet_, well_array_, antigen_position_, well_to_image_):
+    """
+
+    :param sheet_:
+    :param well_array_:
+    :param antigen_position_:
+    :param well_to_image_:
+    :return:
+    """
+    # iterate over all 96 well positions
+    for position, well in np.ndenumerate(well_array_):
+        # extract intensity at antigen_position within this well
+        if well is None:
+            val = -999
```

Comment:
Would it be possible to write Nones instead? It's just easier to distinguish from real data if it's not numbers...

Reply:
Ah yes, I can try Nones no problem.  Will test it first.